### PR TITLE
Adds etcd apiserver client and service account secrets creation

### PR DESCRIPTION
**Pull request by prateekgogia**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: operator/pkg/controllers/etcd/etcd.go [hunk 1]
```diff
@@ -36,14 +36,14 @@ const (
 )
 
 type Controller struct {
-	kubeClient   *kubeprovider.Client
-	certificates *certificates.Provider
+	kubeClient *kubeprovider.Client
+	keypairs   *keypairs.Provider
```
Comment (ellistarn) [inline operator/pkg/controllers/etcd/etcd.go:40]: Name suggestion: `pkiProvider pki.Provider`

File: operator/pkg/controllers/master/kubeconfigs.go [hunk 2]
```diff
@@ -25,18 +25,34 @@ import (
...
+	// kubeconfig certs are signed by master root CA certs, if the root CA is not found return
+	caSecret, err := c.keypairs.GetSecretFromServer(ctx,
+		object.NamespacedName(rootCASecretNameFor(controlPlane.ClusterName()), controlPlane.Namespace))
+	if err != nil {
+		return err
+	}
+	endpoint, err := c.getClusterEndpoint(ctx, object.NamespacedName(
+		controlPlane.ClusterName(), controlPlane.Namespace))
+	if err != nil {
+		return err
```
Comment (ellistarn) [inline operator/pkg/controllers/master/kubeconfigs.go:50]: Can you cut an issue to deep dive error bubbling to make sure that error messages have enough context?

File: operator/pkg/controllers/master/kubeconfigs.go [hunk 3]
```diff
@@ -46,98 +62,108 @@ func (c *Controller) reconcileKubeConfigs(ctx context.Context, controlPlane *v1a
...
+		endpoint: endpoint,
+		auth: &secrets.Request{
+			Name:     kubeControllerManagerSecretNameFor(clusterName),
+			Type:     secrets.KeyWithSignedCert,
+			CASecret: caSecret,
+			Config: &certutil.Config{
+				Usages:     []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
+				CommonName: "system:kube-controller-manager",
+			},
+		}}
```
Comment (ellistarn) [inline operator/pkg/controllers/master/kubeconfigs.go:166]: bit awkward to have the last line of this be }}. If you're going with this style (I'm not really a fan), I'd include the } on line 165, too.

File: operator/pkg/controllers/master/sasecret.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
+/*
```
Comment (ellistarn) [inline operator/pkg/controllers/master/sasecret.go:1]: I think new readers might be confused by the `sa` abbreviation. Consider secret.go if it's the only thing reconciling secrets. Alternatively serviceaccountsecret.go if there are many, but that that point you may want to combine the concepts with secrets.go.

File: operator/pkg/utils/patch/patch.go [hunk 5]
```diff
@@ -54,3 +55,32 @@ func mergePatch(defaultObj, patch, object interface{}) ([]byte, error) {
...
+func parseArgsFor(podSpec *v1.PodSpec) map[string]string {
+	result := map[string]string{}
+	for _, arg := range podSpec.Containers[0].Args {
+		kv := strings.Split(arg, "=")
+		result[kv[0]] = result[kv[1]]
+	}
+	return result
+}
+
+func UnionMapStrings(dest, src map[string]string) map[string]string {
```
Comment (ellistarn) [inline operator/pkg/utils/patch/patch.go:77]: nit: UnionStringMaps

File: operator/pkg/utils/patch/patch.go [hunk 6]
```diff
@@ -54,3 +55,32 @@ func mergePatch(defaultObj, patch, object interface{}) ([]byte, error) {
...
+		merged = append(merged, strings.Join([]string{key, value}, "="))
+	}
+	patch.Containers[0].Args = merged
+	return patch
+}
+
+func parseArgsFor(podSpec *v1.PodSpec) map[string]string {
+	result := map[string]string{}
+	for _, arg := range podSpec.Containers[0].Args {
+		kv := strings.Split(arg, "=")
```
Comment (ellistarn) [inline operator/pkg/utils/patch/patch.go:71]: Cool!

File: operator/pkg/utils/secrets/secrets.go [hunk 7]
```diff
@@ -26,21 +26,41 @@ import (
...
+	CA RequestType = iota + 1
+	KeyPair
+	KeyWithSignedCert
+)
+
 const (
-	SecretCertName = "tls.crt"
-	SecretKeyName  = "tls.key"
+	SecretPrivateKey = "private"
+	SecretPublicKey  = "public"
```
Comment (ellistarn) [inline operator/pkg/utils/secrets/secrets.go:45]: Aren't these names nonstandard? I've always seen tls.crt and tls.key in pki secrets.

File: operator/pkg/controllers/master/kubeconfigs.go (hunk 2 above)
Comment (prateekgogia) [reply operator/pkg/controllers/master/kubeconfigs.go:50]: https://github.com/awslabs/kubernetes-iteration-toolkit/issues/26

File: operator/pkg/controllers/etcd/etcd.go (hunk 1 above)
Comment (prateekgogia) [reply operator/pkg/controllers/etcd/etcd.go:40]: `keypairs` pkg is actually a layer over PKI pkg. It provides logic for checking the API server, if secret exists, if not generate certs/keys.

File: operator/pkg/utils/secrets/secrets.go (hunk 7 above)
Comment (prateekgogia) [reply operator/pkg/utils/secrets/secrets.go:45]: Secret type used is Opaque so don't need to use `tls.crt` and `tls.key`. By using opaque, I can use the same object type for secrets and kubeconfigs in secrets.

**Review by ellistarn — APPROVED** [review]

**Pull request merged**